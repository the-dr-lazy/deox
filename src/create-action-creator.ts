import { createAction, Action, AnyAction } from './create-action'

export type ExactActionCreator<TType extends string, TCallable extends <_T>(...args: any[]) => Action<TType>> = TCallable & {
  type: TType extends AnyAction ? TType['type'] : TType
  toString(): TType extends AnyAction ? TType['type'] : TType
}

export type ActionCreator<T extends AnyAction | string> = {
  (...args: any[]): T extends string ? Action<T> : T
  type: T extends AnyAction ? T['type'] : T
  toString(): T extends AnyAction ? T['type'] : T
}

export type Executor<TType extends string, TCallable extends <_T>(...args: any[]) => Action<TType>> = (
  resolve: <Payload = undefined, Meta = undefined>(
    payload?: Payload,
    meta?: Meta
  ) => Action<TType, Payload, Meta>
) => TCallable;

/**
 * Flux standard action creator factory
 * @example
 * createActionCreator('[Todo] truncate');
 * @example
 * createActionCreator(
 *   '[Todo] fetch rejected',
 *   resolve => (error: Error) => resolve(error)
 * );
 * @example
 * createActionCreator(
 *   '[Todo] fetch rejected',
 *   resolve => (error: Error, meta?: { status: number }) => resolve(error, meta)
 * )
 * @example
 * createActionCreator(
 *   '[Todo] add',
 *   resolve => (name: string, completed = false) => resolve({ name, completed })
 * )
 * @example
 * createActionCreator(
 *   '[Todo] add',
 *   resolve => (name: string, completed = false) => resolve({ name, completed }, 'Meta data of all todos')
 * )
 *
 */
export function createActionCreator<TType extends string>(type: TType): ExactActionCreator<TType, () => Action<TType>>;
export function createActionCreator<
  TType extends string,
  TCallable extends <_T>(...args: any[]) => Action<TType>
>(type: TType, executor: Executor<TType, TCallable>): ExactActionCreator<TType, TCallable>;
export function createActionCreator<
  TType extends string,
  TCallable extends <_T>(...args: any[]) => Action<TType>
>(type: TType, executor: Executor<TType, TCallable> = resolve => (() => resolve()) as TCallable) {
  const callable = executor((payload, meta) =>
    createAction(type, payload!, meta!)
  )

  return Object.assign(callable, {
    type,
    toString() {
      return type
    },
  })
}
