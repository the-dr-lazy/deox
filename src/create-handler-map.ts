import { ActionCreator } from './create-action-creator'
import { AnyAction } from './create-action'
import { getType } from './get-type'
import { Handler } from './types'

export type HandlerMap<
  TPrevState,
  TAction extends AnyAction,
  TNextState extends TPrevState = TPrevState
> = { [type in TAction['type']]: Handler<TPrevState, TAction, TNextState> }

export type InferActionFromHandlerMap<
  THandlerMap extends HandlerMap<any, any>
> = THandlerMap extends HandlerMap<any, infer T> ? T : never

export type InferNextStateFromHandlerMap<
  THandlerMap extends HandlerMap<any, any>
> = THandlerMap extends HandlerMap<any, any, infer T> ? T : never

export type CreateHandlerMap<TPrevState> = (<
  TActionCreator extends ActionCreator<any>,
  TNextState extends TPrevState,
  TAction extends AnyAction = TActionCreator extends (...args: any[]) => infer T
    ? T
    : never
>(
  actionCreators: TActionCreator | TActionCreator[],
  handler: Handler<TPrevState, TAction, TNextState>
) => HandlerMap<TPrevState, TAction, TNextState>) & {
  default: <
    TActionCreator extends ActionCreator<any>,
    TNextState extends TPrevState,
    TAction extends AnyAction = TActionCreator extends (...args: any[]) => infer T
      ? T
      : never
    >(
    handler: Handler<TPrevState, TAction, TNextState>
  ) => { default: Handler<TPrevState, TAction, TNextState> }
}

function handle<
  TActionCreator extends ActionCreator<any>,
  TPrevState,
  TNextState extends TPrevState,
  TAction extends AnyAction = TActionCreator extends (...args: any[]) => infer T
    ? T
    : never
  >(
  actionCreators: TActionCreator | TActionCreator[],
  handler: Handler<TPrevState, TAction, TNextState>
) {
  return (Array.isArray(actionCreators) ? actionCreators : [actionCreators])
    .map(getType)
    .reduce<HandlerMap<TPrevState, TAction, TNextState>>((acc, type) => {
      acc[type] = handler
      return acc
    }, {} as any)
}

handle.default = <
  TActionCreator extends ActionCreator<any>,
  TPrevState,
  TNextState extends TPrevState,
  TAction extends AnyAction = TActionCreator extends (...args: any[]) => infer T
    ? T
    : never
  >(
  handler: Handler<TPrevState, TAction, TNextState>
) => ({ default: handler })

/**
 * Handler map factory
 * @description create an action(s) to reducer map
 * @example
 * createHandlerMap(increment, (state: number) => state + 1)
 * @example
 * createHandlerMap([increment, increase], (state: number) => state + 1)
 * @example
 * createHandlerMap.default((state: number) => state + 1)
 */
export const createHandlerMap = handle as typeof handle & { default: typeof handle.default }


