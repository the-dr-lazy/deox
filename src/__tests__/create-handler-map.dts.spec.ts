import { createActionCreator } from '../create-action-creator'
import { createHandlerMap as handle } from '../create-handler-map'

// @dts-jest:group createHandlerMap

const increment = createActionCreator('[Counter] increment')
const increase = createActionCreator('[Counter] increase')

// @dts-jest:pass:snap
handle(increment, (state: number) => state + 1)

// @dts-jest:pass:snap
handle([increment, increase], (state: number) => state + 1)

// @dts-jest:pass:snap
handle.default((state: number) => state + 1)
