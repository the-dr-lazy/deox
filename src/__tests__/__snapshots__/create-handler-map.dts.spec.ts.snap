// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createHandlerMap handle([increment, increase], (state: number) => state + 1) (type) should match snapshot 1`] = `"HandlerMap<number, { type: \\"[Counter] increment\\"; } | { type: \\"[Counter] increase\\"; }, number>"`;

exports[`createHandlerMap handle(increment, (state: number) => state + 1) (type) should match snapshot 1`] = `"HandlerMap<number, { type: \\"[Counter] increment\\"; }, number>"`;

exports[`createHandlerMap handle.others((state: number) => state + 1) (type) should match snapshot 1`] = `"{ default: Handler<number, any, number>; }"`;
